gui/web_server.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A web server to interact with GUI clients via http or websocket."""
+
+import abc
+
+
+class WebServer(object):
+  """An abstract class of web server to interact with GUI clients.
+
+  Underlying implementation can be based on http or websocket, or support
+  SSL or not.
+  """
+
+  @abc.abstractmethod
+  def Start(self, port, data_callback, error_callback, websocket_url,
```

Comment:
the websocket_url, index_url, html_url, and html_dir seem maybe a bit specific to the WebsocketServer below for an abstract interface

Reply:
I defined WebServer considering we may choose different implementation of websocket or web server, for example, to support python 2 though current websocket needs python 3. Those argument are not websocket specific. They would be valid for other websocket/webserver implementations.